compositor_pipeline/src/pipeline/output/whip/establish_peer_connection.rs

```diff
@@ -0,0 +1,187 @@
+                                    debug!(%err, "Failed to send keyframe request to the encoder.");
+                                };
+                            }
+                        }
+                    } else {
+                        debug!("Failed to read RTCP packets from the sender.");
+                    }
+                }
+            });
+        }
+    }
+
+    let offer = peer_connection
+        .create_offer(None)
+        .await
+        .map_err(WhipError::OfferCreationError)?;
+
+    let endpoint_url = Url::parse(&endpoint_url)
+        .map_err(|e| WhipError::InvalidEndpointUrl(e, endpoint_url.clone()))?;
+
+    info!("[WHIP] endpoint url: {}", endpoint_url);
+
+    let mut header_map = HeaderMap::new();
+    header_map.append("Content-Type", "application/sdp".parse().unwrap());
+
+    let bearer_token = bearer_token.map(Arc::new);
+
+    if let Some(token) = bearer_token.clone() {
+        header_map.append("Authorization", format!("Bearer {token}").parse().unwrap());
+    }
+
+    let response = match client
+        .post(endpoint_url.clone())
+        .headers(header_map)
+        .body(offer.sdp.clone())
+        .send()
+        .await
+    {
+        Ok(res) => res,
+        Err(_) => return Err(WhipError::RequestFailed(Method::POST, endpoint_url)),
+    };
+
+    if response.status() >= StatusCode::BAD_REQUEST {
+        let status = response.status();
+        let answer = response
+            .text()
+            .await
+            .map_err(|e| WhipError::BodyParsingError("sdp offer".to_string(), e))?;
+        return Err(WhipError::BadStatus(status, answer));
+    }
+
+    info!("[WHIP] response: {:?}", &response);
+
+    let location_url_path = response
+        .headers()
+        .get("location")
+        .and_then(|url| url.to_str().ok())
+        .ok_or_else(|| WhipError::MissingLocationHeader)?;
+
+    let scheme = endpoint_url.scheme();
+    let host = endpoint_url
+        .host_str()
+        .ok_or_else(|| WhipError::MissingHost)?;
+
+    let port = endpoint_url.port().ok_or_else(|| WhipError::MissingPort)?;
+
+    let formatted_url = format!("{}://{}:{}{}", scheme, host, port, location_url_path);
+
+    let location_url = Url::try_from(formatted_url.as_str())
+        .map_err(|e| WhipError::InvalidEndpointUrl(e, formatted_url))?;
+
+    peer_connection
+        .set_local_description(offer)
+        .await
+        .map_err(WhipError::LocalDescriptionError)?;
+
+    let answer = response
+        .text()
+        .await
+        .map_err(|e| WhipError::BodyParsingError("sdp offer".to_string(), e))?;
+
+    let rtc_answer =
+        RTCSessionDescription::answer(answer).map_err(WhipError::RTCSessionDescriptionError)?;
+
+    peer_connection
+        .set_remote_description(rtc_answer)
+        .await
+        .map_err(WhipError::RemoteDescriptionError)?;
+
+    let client = Arc::new(client);
+
+    let location1: Url = location_url.clone();
```

Comment:
do not add numbers to variable names, if you have to create a clone to move it somewhere use `_clone`